fdbserver/GlobalTagThrottler.actor.cpp

```diff
@@ -27,197 +27,495 @@
+			return it->second;
+		}
+	}
 
-		Optional<double> getReadTPSLimit() const {
-			if (totalReadCostRate.smoothTotal() > 0) {
-				return quota.get().totalReadQuota * transactionCounter.smoothRate() / totalReadCostRate.smoothTotal();
+	class ThroughputCounters {
+		Smoother readCost;
+		Smoother writeCost;
+
+	public:
+		ThroughputCounters()
+		  : readCost(SERVER_KNOBS->GLOBAL_TAG_THROTTLING_FOLDING_TIME),
+		    writeCost(SERVER_KNOBS->GLOBAL_TAG_THROTTLING_FOLDING_TIME) {}
+
+		// Returns difference between new and current rates
+		double updateCost(double newCost, OpType opType) {
+			if (opType == OpType::READ) {
+				auto const currentReadCost = readCost.getTotal();
+				readCost.setTotal(newCost);
+				return newCost - currentReadCost;
 			} else {
-				return {};
+				auto const currentWriteCost = writeCost.getTotal();
+				writeCost.setTotal(newCost);
+				return newCost - currentWriteCost;
 			}
 		}
 
-		Optional<double> getWriteTPSLimit() const {
-			if (totalWriteCostRate.smoothTotal() > 0) {
-				return quota.get().totalWriteQuota * transactionCounter.smoothRate() / totalWriteCostRate.smoothTotal();
+		double getCost(OpType opType) const {
+			if (opType == OpType::READ) {
+				return readCost.smoothTotal();
 			} else {
-				return {};
+				return writeCost.smoothTotal();
 			}
 		}
+	};
+
+	// Track various statistics per tag, aggregated across all storage servers
+	class PerTagStatistics {
+		Optional<ThrottleApi::TagQuotaValue> quota;
+		Smoother transactionCounter;
+		Smoother perClientRate;
 
 	public:
-		QuotaAndCounters()
-		  : totalReadCostRate(SERVER_KNOBS->GLOBAL_TAG_THROTTLING_FOLDING_TIME),
-		    totalWriteCostRate(SERVER_KNOBS->GLOBAL_TAG_THROTTLING_FOLDING_TIME),
-		    transactionCounter(SERVER_KNOBS->GLOBAL_TAG_THROTTLING_FOLDING_TIME),
+		explicit PerTagStatistics()
+		  : transactionCounter(SERVER_KNOBS->GLOBAL_TAG_THROTTLING_FOLDING_TIME),
 		    perClientRate(SERVER_KNOBS->GLOBAL_TAG_THROTTLING_FOLDING_TIME) {}
 
-		void setQuota(ThrottleApi::TagQuotaValue const& quota) { this->quota = quota; }
+		Optional<ThrottleApi::TagQuotaValue> getQuota() const { return quota; }
+
+		void setQuota(ThrottleApi::TagQuotaValue quota) { this->quota = quota; }
+
+		void clearQuota() { quota = {}; }
+
+		void addTransactions(int count) { transactionCounter.addDelta(count); }
+
+		double getTransactionRate() const { return transactionCounter.smoothRate(); }
 
-		void updateReadCostRate(UID ssId, double newReadCostRate) {
-			auto& currentReadCostRate = ssToReadCostRate[ssId];
-			auto diff = newReadCostRate - currentReadCostRate;
-			currentReadCostRate += diff;
-			totalReadCostRate.addDelta(diff);
+		ClientTagThrottleLimits updateAndGetPerClientLimit(Optional<double> targetTps) {
+			auto newPerClientRate =
+			    std::max(SERVER_KNOBS->GLOBAL_TAG_THROTTLING_MIN_RATE,
+			             std::min(targetTps.get(),
+			                      (targetTps.get() / transactionCounter.smoothRate()) * perClientRate.smoothTotal()));
```

Comment:
What if `targetTPS.present() == false`? If it's impossible, how about changing the argument type to double?